Added more information on token auth workflow. (#1626)
pulp/pulp_container: staging_docs/admin/learn/authentication.md

```

Restart Pulp services in order to reload the updated settings. Pulp will fetch a domain for the token
server and will initialize all handlers according to that.

!!! note

    Standard container tooling clients like podman, skopeo and docker handle token authentication calls on
    the behalf of the user during pull and push operations, however, if you would like to debug registry
    access or implement manual registry API calls, here are few examples on how token authentication works
    behind the scene.

Access the root registry endpoint to check if the token authentication was successfully configured and enabled:

```bash
    $ http 'https://puffy.example.com/v2/'

    HTTP/1.1 401 Unauthorized
    Allow: GET, HEAD, OPTIONS
    Connection: close
    Content-Length: 58
    Content-Type: application/json
    Date: Mon, 13 Jul 2020 09:56:54 GMT
    Docker-Distribution-Api-Version: registry/2.0
    Server: gunicorn/20.0.4
    Vary: Accept
    WWW-Authenticate: Bearer realm="https://puffy.example.com/token/",service="puffy.example.com"
    X-Frame-Options: SAMEORIGIN

    {
        "detail": "Authentication credentials were not provided."
    }
```

> Since the request was not authenticated, registry returned 401 and in the authentication header there are

Reviewer: ```suggestion
Since the request was not authenticated, the registry returned 401 and in the authentication header there are
```

details on how to authenticate. This time send the request with the specified realm and service:

```bash
$ http https://puffy.example.com/token/?service=puffy.example.com

HTTP/1.1 200 OK
Allow: GET, HEAD, OPTIONS
Connection: close
Content-Length: 609
Content-Type: application/json
Date: Mon, 13 Jul 2020 09:57:25 GMT
Server: gunicorn/20.0.4
Vary: Accept, Cookie
X-Frame-Options: SAMEORIGIN

{   
    "access_token": <TOKEN>,
    "expires_in": 300,
    "issued_at": "2020-07-13T09:57:25.601760Z",
    "token": <TOKEN>
}
```

The token was generated to access registry root endpoint.

```bash
$ http --auth-type=jwt --auth=<TOKEN> https://puffy.example.com/v2/

HTTP/1.1 200 OK
Allow: GET, HEAD, OPTIONS
Connection: close
Content-Length: 2
Content-Type: application/json
Date: Mon, 13 Jul 2020 09:58:40 GMT
Docker-Distribution-Api-Version: registry/2.0
Server: gunicorn/20.0.4
Vary: Accept
X-Frame-Options: SAMEORIGIN

{}
```

> In order to access other registry endpoints, like manifests, blobs and tags the scope needs to be provided
> as well. In this context it will be a specific repository with the according action pull or push.

Reviewer: ```suggestion
In order to access other registry endpoints, like manifests, blobs and tags, the scope needs to be provided
as well. In this context, it will be a specific repository with the according action pull or push.
```

```bash
$ http https://puffy.example.com/v2/library/azure/tags/list

> HTTP/1.1 401 Unauthorized

Reviewer: Do you mind separating the request and response snippets, like we have here: https://github.com/pulp/pulp_container/blob/32f7cabd1c3fd23b4650ad1cc09681a9c43aa62a/staging_docs/admin/learn/rbac.md?plain=1#L44-L59?

Connection: keep-alive
Content-Length: 1311
Content-Type: application/json
Correlation-ID: 611f0ce122a54509b348ff54d5030c80
Cross-Origin-Opener-Policy: same-origin
Date: Fri, 17 May 2024 09:20:29 GMT
Referrer-Policy: same-origin
Server: nginx
Strict-Transport-Security: max-age=15768000
Vary: Accept
X-Content-Type-Options: nosniff
X-Frame-Options: DENY

{
    "access_token": <TOKEN>,
    "expires_in": 300,
    "issued_at": "2024-05-17T09:20:29.660984Z",
    "token": <TOKEN>
}

$ http https://puffy.example.com/v2/library/azure/tags/list --auth-type=jwt --auth=<TOKEN>
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Correlation-ID
Allow: GET, HEAD, OPTIONS
Connection: keep-alive
Content-Length: 42
Content-Type: application/json
Correlation-ID: b944d8b6b1fc474eada8531776a2609b
Cross-Origin-Opener-Policy: same-origin
Date: Fri, 17 May 2024 09:33:04 GMT
Docker-Distribution-Api-Version: registry/2.0
Referrer-Policy: same-origin
Server: nginx
Strict-Transport-Security: max-age=15768000
Vary: Accept
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Registry-Supports-Signatures: 1

{
    "name": "library/azure",
    "tags": [
        "latest"
    ]
}
```

!!! note

>     If the repository is private, in order to pull content from it, one would need to provide basic auth
>     credentials while requesting the token. If no basic auth is provided, then the generated token is for

Reviewer: ```suggestion
    If the repository is private, one needs to provide basic auth
    credentials while requesting the token in order to pull content from it. If no basic auth is provided, then the generated token is for
```

    the anonymous user.

>     In order to push content to the registery a user always needs to be autenticated.

Reviewer: ```suggestion
    In order to push content to the registry, a user always needs to be authenticated.
```

```bash
$ http 'https://puffy.example.com/token/?service=puffy.example.com&scope=repository:private/azure:pull' --auth-type=basic --auth=alice:wonderland
```

!!! note

>     Some registry endpoints like ``_catalog`` endpoint are not opened to anonymous/not logged in users and
>     require credentials provided during the token request. Anonymous token will still lead to 401 insufficient scope.

Reviewer: ```suggestion
    Some registry endpoints, like ``_catalog`` endpoint, are not opened to anonymous/not logged-in users and
    require credentials provided during the token request. Anonymous tokens will still lead to HTTP 401 insufficient scope errors.
```

```bash
http 'https://puffy.example.com/token/?service=puffy.example.com&scope=registry:catalog:*' --auth-type=basic --auth=alice:wonderland
```

This token embeds permissions that allow to see only those repositories that Alice has access to. 

```bash
$ http https://puffy.example.com/v2/_catalog --auth-type=jwt --auth=<TOKEN>
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Correlation-ID
Allow: GET, HEAD, OPTIONS
Connection: keep-alive
Content-Length: 63
Content-Type: application/json
Correlation-ID: f4ea6c26bcb7462099c0d5757178f7ca
Cross-Origin-Opener-Policy: same-origin
Date: Fri, 17 May 2024 09:57:11 GMT
Docker-Distribution-Api-Version: registry/2.0
Referrer-Policy: same-origin
Server: nginx
Strict-Transport-Security: max-age=15768000
Vary: Accept
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Registry-Supports-Signatures: 1

{
    "repositories": [
        "alice/azure",
        "alice/openstack-cron"
    ]
}
```

> If you are still unsure why you are geting 401 with the generated token, paste its payload to https://jwt.io/
> and it will help identify whether the token contains the needed access.

Reviewer: ```suggestion
If you are still unsure why you are getting HTTP 401 errors with a generated token, paste its payload to https://jwt.io/
and it will help you identify whether the token contains the needed access.
```

Below is provided an example of the settings file:

```python
TOKEN_SERVER = "https://puffy.example.com/token/"
TOKEN_SIGNATURE_ALGORITHM = 'ES256'
PUBLIC_KEY_PATH = '/tmp/public_key.pem'
PRIVATE_KEY_PATH = '/tmp/private_key.pem'
```

Restart Pulp services in order to reload the updated settings. Pulp will fetch a domain for the token
server and will initialize all handlers according to that.

!!! note

>     Standard container tooling clients like podman, skopeo and docker handle token authentication calls on
>     the behalf of the user during pull and push operations, however, if you would like to debug registry
>     access or implement manual registry API calls, here are few examples on how token authentication works
>     behind the scene.

Reviewer: ```suggestion
    Standard container tooling clients like podman, skopeo and docker handle token authentication calls on
    behalf of the user during pull and push operations. However, if you would like to debug registry
    access or implement manual registry API calls, here are a few examples of how the token authentication works
    behind the scene.
```
